Define Keyset ID V2 (#182)
cashubtc/nuts: 02.md

### Keyset ID

A keyset `id` is an identifier for a specific keyset. It can be derived by anyone who knows the set of public keys of a mint. Wallets **MAY** compute the keyset `id` for a given keyset by themselves to confirm that the mint is supplying the correct keyset ID (see below).

The keyset `id` is in each `Proof` so it can be used by wallets to identify which mint and keyset it was generated from. The keyset field `id` is also present in the `BlindedMessages` sent to the mint and `BlindSignatures` returned from the mint (see [NUT-00][00]).

> To save space, a `Token`, as definied in [NUT-00][00], **SHOULD** contain an abbreviated version of the keyset `id` in the `Proof`. The length of the abbreviated `id` **MUST** be at least two bytes (i.e., the version byte and first byte of the hash). However, when the token is serialized, the chosen `id` length **MUST** not be ambiguous within the mint's existing keyset `id`s. A wallet **MUST** resolve the abbreviated keyset `id` to the full `id`. If the abbreviated `id` is ambiguous (i.e., multiple keyset `id`s are resolvable), the wallet **MUST** error. The full keyset `id` is only needed when the wallet interacts with the mint.

Reviewer: What happens to the tokens when a mint generates a new keyset that collides with the already emitted tokens?

Let's say the old keyset id is `01cb2343....`, the tokens use `01cb` in the `i` field and then a new keyset is created which has id `01cb948...`. The chance of this happening is 1:256 (around 0.4%) - if you generate more keysets, the chance is significantly higher.

### Keyset ID

A keyset `id` is an identifier for a specific keyset. It can be derived by anyone who knows the set of public keys of a mint. Wallets **MAY** compute the keyset `id` for a given keyset by themselves to confirm that the mint is supplying the correct keyset ID (see below).

The keyset `id` is in each `Proof` so it can be used by wallets to identify which mint and keyset it was generated from. The keyset field `id` is also present in the `BlindedMessages` sent to the mint and `BlindSignatures` returned from the mint (see [NUT-00][00]).

> To save space, a `Token`, as defined in [NUT-00][00], **SHOULD** contain an abbreviated version of the keyset `id` in the `Proof`, (the `s_id`). The length of the `s_id` **MUST** be eight bytes (i.e., the version byte and the first seven bytes of the hash: `id_bytes[:8]` or `id_hex[:16]`). A wallet **MUST** resolve the abbreviated keyset `id` to the full `id`. If the abbreviated `s_id` is ambiguous (i.e., multiple keyset `id`s are resolvable), the wallet **MUST** error. The full keyset `id` is only needed when the wallet interacts with the mint.

Reviewer 2: ```suggestion
### Short ID

The short keyset ID (`s_id`) is defined as the abbreviated representation of the full Keyset ID. The length of the `s_id` is eight bytes (i.e., the version byte and the first seven bytes of the hash: `id_bytes[:8]` or `id_hex[:16]`).

The `s_id` is only relevant for wallets, when encoding and decoding `Tokens`. To save space, a `Token`, as defined in [NUT-00][00], **SHOULD** use the short keyset ID `s_id` in the `Proof`. The recommended exception to that is when the `Proofs` reference a keyset where the `s_id` is ambiguous on that mint; in that case, the sending wallet **SHOULD** instead use the full keyset ID. Receiving wallets **MUST** therefore be able to decode both `Tokens` with a short keyset ID, as well as those with a full keyset ID.

When receiving a `Token` with a `s_id`, a wallet **MUST** resolve the abbreviated `s_id` to the full `id`. If the `s_id` is ambiguous (i.e., multiple keyset `id`s are resolvable on the same mint), the wallet **MUST** error. Otherwise, if the `Token` is successfully parsed, the wallet **SHOULD** save the full-length keyset `id` with proofs in its database.

The mint is unaware of the `s_id`. All API calls exposed by the mint use the full keyset ID.
```

As discussed, the ambiguity mentioned above ( https://github.com/cashubtc/nuts/pull/182#issuecomment-2773223419 ) is intentional and the spec should allow for both short and long Keyset IDs in the `Token`.

@davidcaseria I re-worded the paragraph that originally confused me. It now spells out in more detail who should use the `s_id` when.

> To save space, a `Token`, as definied in [NUT-00][00], **SHOULD** contain an abbreviated version of the keyset `id` in the `Proof`. The length of the abbreviated `id` **MUST** be at least eight bytes (i.e., the version byte and first byte of the hash). However, when the token is serialized, the chosen `id` length **MUST** not be ambiguous within the mint's existing keyset `id`s. A wallet **MUST** resolve the abbreviated keyset `id` to the full `id`. If the abbreviated `id` is ambiguous (i.e., multiple keyset `id`s are resolvable), the wallet **MUST** error. The full keyset `id` is only needed when the wallet interacts with the mint.

Reviewer 2: > ... the abbreviated `id` **MUST** be at least eight bytes (i.e., the version byte and first byte of the hash)

8 != 1 + 1

Author: Oops. Thanks!

> To save space, a `Token`, as defined in [NUT-00][00], **SHOULD** contain an abbreviated version of the keyset `id` in the `Proof`. The length of the abbreviated `id` **MUST** be at least eight bytes (i.e., the version byte and the first seven bytes of the hash). However, while unlikely, when the token is serialized, the chosen `id` length **MUST** not be ambiguous within the mint's existing keyset `id`s. A wallet **MUST** resolve the abbreviated keyset `id` to the full `id`. If the abbreviated `id` is ambiguous (i.e., multiple keyset `id`s are resolvable), the wallet **MUST** error. The full keyset `id` is only needed when the wallet interacts with the mint.

Reviewer 3: This seems to vague and leaves room for interpretation. It could simply say that the token must include the first 8 bytes / 16 hex characters of the keyset ID it originates from.

Reviewer 3: We could make it more explicit and call it `s_id` which is `id_hex[:16]` or `id_bytes[:8]`.

### Keyset ID

A keyset `id` is an identifier for a specific keyset. It can be derived by anyone who knows the set of public keys of a mint. Wallets **SHOULD** compute the keyset `id` for a given keyset themselves to confirm that the mint is supplying the correct keyset ID (see below).

The keyset `id` is in each `Proof` so it can be used by wallets to identify which mint and keyset it was generated from. The keyset field `id` is also present in the `BlindedMessages` sent to the mint and `BlindSignatures` returned from the mint (see [NUT-00][00]).

### Short ID

The short keyset ID (`s_id`) is defined as the abbreviated representation of the full Keyset ID. The length of the `s_id` is eight bytes (i.e., the version byte and the first seven bytes of the hash: `id_bytes[:8]` or `id_hex[:16]`).

> The `s_id` is only relevant for wallets, when encoding and decoding `Tokens`. To save space, a `Token`, as defined in [NUT-00][00], **SHOULD** use the short keyset ID `s_id` in the `Proof`. The recommended exception to that is when the `Proofs` reference a keyset where the `s_id` is ambiguous on that mint; in that case, the sending wallet **SHOULD** instead use the full keyset ID. Receiving wallets **MUST** therefore be able to decode both `Tokens` with a short keyset ID, as well as those with a full keyset ID.

Reviewer 3: ```suggestion
Wallets use `s_id` to save space when encoding and decoding `Tokens` as defined in [NUT-00][00]. The recommended exception to that is when the `s_id` is ambiguous; in that case, the sending wallet **SHOULD** instead use the full keyset ID. Receiving wallets **MUST** be able to decode both `Tokens` with a short keyset ID, as well as those with a full keyset ID.
```

The `s_id` is only relevant for wallets, when encoding and decoding `Tokens`. To save space, a `Token`, as defined in [NUT-00][00], **SHOULD** use the short keyset ID `s_id` in the `Proof`. The recommended exception to that is when the `Proofs` reference a keyset where the `s_id` is ambiguous on that mint; in that case, the sending wallet **SHOULD** instead use the full keyset ID. Receiving wallets **MUST** therefore be able to decode both `Tokens` with a short keyset ID, as well as those with a full keyset ID.

When receiving a `Token` with a `s_id`, a wallet **MUST** resolve the abbreviated `s_id` to the full `id`. If the `s_id` is ambiguous (i.e., multiple keyset `id`s are resolvable on the same mint), the wallet **MUST** error. Otherwise, if the `Token` is successfully parsed, the wallet **SHOULD** save the full-length keyset `id` with proofs in its database.

The mint is unaware of the `s_id`. All API calls exposed by the mint use the full keyset ID.

> 
> A Wallet **SHOULD** save the full-length keyset `id` with proofs in its database.

Reviewer 2: ```suggestion
```

NIT: Duplicate line, it's already included in the last part of the paragraph on L27 above.

### Deriving the keyset ID

#### Keyset ID V2

Keyset IDs are 33 byte hex strings with a version byte (two hexadecimal characters). The currently used version byte is `01`.

Keyset IDs are derived from public data. To derive the keyset ID of a keyset, execute the following steps:

```
1 - sort public keys by their amount in ascending order
2 - concatenate all public keys to a single byte array
3 - add the lowercase UTF8-encoded unit string to the byte array (e.g. "unit:sat")
4 - If a final expiration is specified, add the UTF8-encoded string (e.g "final_expiry:1896187313")
4 - HASH_SHA256 the concatenated byte array

> 5 - prefix it with a keyset ID version byte "01"

Reviewer 4: The keyset ID derivation doesn't protect/guarantee the amounts have not been tampered with, only that the amounts are still ordered the same.

What stops a mint from changing the amounts bound to each key to devalue and "inflate" away their liabilities?

eg: 1, 2, 4, 8, 16, 32, 64

could become: 1, 2, 3, 4, 8, 16, 32
or maybe even: 1, 1, 1, 1, 1, 1, 1

Perhaps we should concat all the amounts in order too... or maybe just adding the sum of key amounts would be enough to reduce the risk of serious damage?

Reviewer 3: > f we will be changing this, I would suggest to include my suggestion which was dismissed earlier: [#182 (comment)](https://github.com/cashubtc/nuts/pull/182#discussion_r2306837257)

yes! we need to add separators if we append the amounts as well, let's combine both changes

### Deriving the keyset ID

#### Keyset ID

Keyset IDs have a version byte (two hexadecimal characters). The currently used version byte is `01`.

The mint and the wallets of its users can derive a keyset ID from the keyset of the mint. The keyset ID is a lower-case hex string. To derive the keyset ID of a keyset, execute the following steps:

```
1 - sort public keys by their amount in ascending order
2 - concatenate all public keys to one byte array
3 - add the unit string to the byte array (e.g. "unit:sat")
4 - HASH_SHA256 the concatenated byte array
5 - prefix it with a keyset ID version byte
```

An example implementation in Python:

```python
def derive_keyset_id(keys: Dict[int, PublicKey]) -> str:
    sorted_keys = dict(sorted(keys.items()))
    keyset_id_bytes = b"".join([p.serialize() for p in sorted_keys.values()])

>     keyset_id_bytes += b"unit:sat"

Reviewer 2: Once optimization could be to change `b"unit:sat"` -> `b"u:sat"` or even just `b"sat"`. Might also be worth specifying the unit should be lowercased first.

Author: I don't think we have specified the units as case-insensitive yet. We should probably also do that if we include this change, right?

Reviewer 2: Since the keyset ID commits to the unit, then it would make sense IMO.

If a client derives the keyset ID using a different capitalization, or if the mint changes the capitalization later on, the keyset ID would be invalid.

Author: @callebtc what do you think about units being case-insensitive and hashing the lowercase unit value?

Reviewer 5: We should specify we hash the lower case.

### Deriving the keyset ID

> [!CRITICAL]
> Wallet implementations should reject any attempt at importing new keysets which IDs
> collide with any of the previously added keysets.

#### Keyset ID

Keyset IDs have a version byte (two hexadecimal characters). The currently used version byte is `01`.

The mint and the wallets of its users can derive a keyset ID from the keyset of the mint. The keyset ID is a lower-case hex string. To derive the keyset ID of a keyset, execute the following steps:

```
1 - sort public keys by their amount in ascending order
2 - concatenate all public keys to one byte array
3 - add the lowercase unit string to the byte array (e.g. "unit:sat")
4 - If a final expiration is specified, convert it into a radix-10 string and add it (e.g "final_expiry:1896187313")
4 - HASH_SHA256 the concatenated byte array
5 - prefix it with a keyset ID version byte
```

An example implementation in Python:

```python
def derive_keyset_id(keys: Dict[int, PublicKey]) -> str:

>     sorted_keys = dict(sorted(keys.items()))
>     keyset_id_bytes = b"".join([p.serialize() for p in sorted_keys.values()])
>     keyset_id_bytes += b"unit:sat"
>     keyset_id_bytes += b"final_expiry:"+str(1896187313).encode("utf-8")

Reviewer: can we add delimiters here?

they won't change the size of the result and makes it so much better for debugging implementations

example

```suggestion
    sorted_keys = dict(sorted(keys.items()))
    keyset_id_bytes = b",".join([p.serialize() for p in sorted_keys.values()])
    keyset_id_bytes += b"|unit:sat"
    keyset_id_bytes += b"|final_expiry:"+str(1896187313).encode("utf-8")
```

```python
def derive_keyset_id(keys: Dict[int, PublicKey]) -> str:
    sorted_keys = dict(sorted(keys.items()))
    keyset_id_bytes = b"".join([p.serialize() for p in sorted_keys.values()])

>     keyset_id_bytes += b"unit:sat"

Reviewer: I am wondering ... since we already depend for CBOR (for Token v4), maybe we can use CBOR here too instead of this custom serialization?

Author: What would the structure of the CBOR be?

      "unit": <str>,
      "active": <bool>,
      "input_fee_ppk": <int|null>,

>       "final_expiry": <int|null>

Reviewer 3: that's how we usually indicate that a value is optional through the spec.

is it an equivalent "json" to omit a key if the value is `null`? i.e.`{a: 1, b: null} == {a: 1} `?

      "unit": <str>,
      "active": <bool>,
      "input_fee_ppk": <int|null>,

>       "final_expiry": <int|null>

Reviewer 6: When I first wrote this what I imagined could happen is at some point we'll want other expiry fields (for example `active_expiry` to signal when this keyset is expected to be made inactive). `final_expiry` is stronger and signals a potential deletion from the DB.

{
  "keysets": [

>     {

Reviewer 3: I think it's fine, since we're deprecating v1 with this.

Request of `Alice`:

> We request the keys for the keyset `01c9c20fb8b348b389e296227c6cc7a63f77354b7388c720dbba6218f720f9b785`.

Reviewer 6: What happens if a wallet that does not yet support V2 keysets makes a `GET v1/keys/{8-byte-keyset-id}` request?

Author: A request with a keyset ID prefix `00` should still be supported. A wallet that doesn't support keyset IDs with a prefix of `01` shouldn't get to where they are making this request since the keyset ID parsing will have already failed if the wallet is checking the version.

If a wallet doesn't check the version and sends a request for an 8-byte keyset ID from a token, we can consider having the mint resolve it for the wallet from the URL. But that feels like an optional requirement for the mint.

Does that address your question?
